src/injected/web/gm-api.js

```diff
@@ -152,20 +154,21 @@ export function makeGmApi() {
     GM_xmlhttpRequest(opts) {
       return onRequestCreate(opts, this.id);
     },
-    GM_addStyle(css) {
-      const id = bridge.sendSync('AddStyle', css);
-      const el = document::getElementById(id);
-      // Mock a Promise without the need for polyfill
-      // It's not actually necessary because DOM messaging is synchronous
-      // but we keep it for compatibility with VM's 2017-2019 behavior
-      // https://github.com/violentmonkey/violentmonkey/issues/217
-      el.then = callback => {
-        // prevent infinite resolve loop
-        delete el.then;
-        callback(el);
-      };
-      return el;
-    },
+    /**
+     * Bypasses site's CSP for inline `style`, `link`, and `script`.
+     * @param {Node} [parent]
+     * @param {string} tag
+     * @param {Object} [attributes]
+     * @returns {HTMLElement}
+     */
+    GM_addElement: (parent, tag, attributes) => (
+      typeof parent === 'string'
+        ? addElement(undefined, parent, tag)
+        : addElement(parent, tag, attributes)
+    ),
+    GM_addStyle: css => (
+      addElement(undefined, 'style', { textContent: css, id: getUniqId('VMst') })
```

Comment:
@gera2ld, now that AddElement is using parentIndex, I wonder if we can remove `id: getUniqId('VMst')` ?

Reply:
Yeah, I think it's useless now.

Reply:
After a lot of attempts to handle ShadowDOM (all were too complicated and wouldn't work if `parent` is a detached node not in DOM), I reverted to VM's old approach with a temporary `id` so the style's id will stay to avoid breaking something.